### PR TITLE
Fix EVP_PKEY_asn1_find() and EVP_PKEY_asn1_find_str() to always look …

### PR DESCRIPTION
**Pull request by anhu**
…in engines

Ran `make tests` and make `doc-nits` and everything was fine.

### CODE REVIEW

**Comment by anhu** [comment]
please note that I am anth_hu@hotmail.com and so you clear the CLA hold.

**Comment by richsalz** [comment]
you can change the author on the commit (git commit --am --au .....) or perhaps send email to the CLA email address so that they can update the database.

**Comment by anhu** [comment]
ugh...didn't notice that email address in the Author field. thank you.

**Comment by anhu** [comment]
please note commit how has an author that is in your CLA list. please clear the flag.

**Comment by levitte** [comment]
This is a semantic change that I'm not too sure about. What's the use case that got you to make this change?

**Comment by anhu** [comment]
I'm building a version of openssl where algorithms are not implemented in openssl but in an engine so its critical that all calls to `EVP_PKEY_asn1_find()` looks for implementations in engines.  I know what you're thinking, just make sure all calls to `EVP_PKEY_asn1_find()` send in a tmp_eng ptr.  

But I think this change has much more value to the wider community because a developer might try to disable the openssl implementation of an algorithm and only use their own engine implementation.

**Comment by richsalz** [comment]
@anhu -- the flag is done automatically, such as when you force-push :)

**Comment by anhu** [comment]
> @anhu -- the flag is done automatically, such as when you force-push :)

Ah....thank you for letting me know.

**Comment by anhu** [comment]
> This is a semantic change that I'm not so sure about. What's the use case that got you to make this change?

Also, I can't think of a use case where checking the engines would be a bad idea.

**Comment by anhu** [comment]
All the following cases trigger not checking for engines: 
```
ssl/ssl_ciph.c:326:    ameth = EVP_PKEY_asn1_find_str(NULL, pkey_name, -1);
crypto/x509/x509_set.c:211:        const EVP_PKEY_ASN1_METHOD *ameth = EVP_PKEY_asn1_find(NULL, pknid);
crypto/x509/t_x509.c:318:            ameth = EVP_PKEY_asn1_find(NULL, pkey_nid);
crypto/pem/pem_lib.c:144:            ameth = EVP_PKEY_asn1_find_str(NULL, nm, slen);
crypto/store/loader_file.c:424:                && (ameth = EVP_PKEY_asn1_find_str(NULL, pem_name,
```
Why not future proof it?

**Comment by levitte** [comment]
> Why not future proof it?

You may not have followed recent development...  ENGINE is considered legacy with upcoming v3.0, and a new concept called "provider" replaces it.  See the [recent blog post](https://www.openssl.org/blog/blog/2019/11/07/3.0-update/)...

**Comment by anhu** [comment]
Oh, yes.  But from what I understand, the engine API will still remain for backwards compatibility, correct? As well, 1.1.1 is LTS so this change could potentially help the community until  2023-09-11.

**Comment by mattcaswell** [comment]
> Oh, yes. But from what I understand, the engine API will still remain for backwards compatibility, correct?

It will remain for now but be deprecated. In some future version it will be removed. We will strongly encourage everyone to move to providers.

**Comment by anhu** [comment]
Hi, 

Might I propose an alternative? All the following calls could have engine pointers that would do nothing.  Would that be a more preferable solution?  While I understand that engines are going out, I still they they will be around on the order of years...no? 

```
ssl/ssl_ciph.c:326:    ameth = EVP_PKEY_asn1_find_str(NULL, pkey_name, -1);
crypto/x509/x509_set.c:211:        const EVP_PKEY_ASN1_METHOD *ameth = EVP_PKEY_asn1_find(NULL, pknid);
crypto/x509/t_x509.c:318:            ameth = EVP_PKEY_asn1_find(NULL, pkey_nid);
crypto/pem/pem_lib.c:144:            ameth = EVP_PKEY_asn1_find_str(NULL, nm, slen);
crypto/store/loader_file.c:424:                && (ameth = EVP_PKEY_asn1_find_str(NULL, pem_name,
```

**Comment by anhu** [comment]
Anyone have any objections? 

Upon closer inspection, this one:

```
ssl/ssl_ciph.c:326:    ameth = EVP_PKEY_asn1_find_str(NULL, pkey_name, -1);
```
...does not need modification because there is another version of that function that is implemented if engines are enabled. (coincidentally, that other version could be removed if this change were to be accepted. :) )

**Comment by anhu** [comment]
> It will remain for now but be deprecated. In some future version it will be removed. 
I would assume not until 1.1.1 is EOL.  Is that a bad assumption?

**Comment by anhu** [comment]
Hi, 
It seems that there isn't very much enthusiasm for this PR.  Should I close?

File: crypto/asn1/ameth_lib.c [hunk 1]
```diff
@@ -71,33 +71,36 @@ static const EVP_PKEY_ASN1_METHOD *pkey_asn1_find(int type)
...
-    if (pe) {
+
 #ifndef OPENSSL_NO_ENGINE
-        ENGINE *e;
-        /* type will contain the final unaliased type */
-        e = ENGINE_get_pkey_asn1_meth_engine(type);
-        if (e) {
+    /* type will contain the final unaliased type */
+    e = ENGINE_get_pkey_asn1_meth_engine(type);
+    if (e) {
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: Please note that our coding style mandates `if (e != NULL)` for NULL pointer checks.

_(I know it was this way already, but please fix it while you're at it)_

File: crypto/asn1/ameth_lib.c [hunk 2]
```diff
@@ -71,33 +71,36 @@ static const EVP_PKEY_ASN1_METHOD *pkey_asn1_find(int type)
...
+        if (pe)
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: same here

File: crypto/asn1/ameth_lib.c [hunk 3]
```diff
@@ -71,33 +71,36 @@ static const EVP_PKEY_ASN1_METHOD *pkey_asn1_find(int type)
...
             *pe = e;
-            return ENGINE_get_pkey_asn1_meth(e, type);
-        }
+        return ENGINE_get_pkey_asn1_meth(e, type);
+    }
 #endif
+    if (pe)
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: and here

File: crypto/asn1/ameth_lib.c [hunk 4]
```diff
@@ -120,12 +125,17 @@ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **pe,
             if (!ENGINE_init(e))
                 ameth = NULL;
             ENGINE_free(e);
-            *pe = e;
+            if (pe) {
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: here

File: crypto/asn1/ameth_lib.c [hunk 5]
```diff
@@ -120,12 +125,17 @@ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **pe,
...
+                *pe = NULL;
+                if (ameth)
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: here

File: crypto/asn1/ameth_lib.c [hunk 6]
```diff
@@ -120,12 +125,17 @@ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **pe,
...
+                    *pe = e;
+            }
             return ameth;
         }
 #endif
+    if (pe)
```
Comment (mspncp) [inline crypto/asn1/ameth_lib.c]: here

File: crypto/asn1/ameth_lib.c (hunk 1 above)
Comment (anhu) [reply crypto/asn1/ameth_lib.c]: OH!  For sure!  `!= NULL` has always been my preference, but I put pre-existing style conformance as the top priority.  Will happily make these changes soon.

**Comment by levitte** [comment]
This PR will simply not do, because it will leak ENGINE references.  It's all in this line:

https://github.com/openssl/openssl/blob/3478a2102e8bfc63f20e22826ab39b407ff43f62/crypto/asn1/ameth_lib.c#L93

`ENGINE_get_pkey_asn1_meth_engine()` returns not just an ENGINE pointer, but a functional reference, which means that this ENGINE instance has been initialized for this use.  At some point, someone must call `ENGINE_finish()` on this pointer, or that ENGINE's functional reference count will not be decremented.

This is the whole reason why it won't look at engines unless `pe` is non-NULL.

So, the better fix is actually to make sure all calls are with a non-NULL `pe`.

**Comment by anhu** [comment]
Sorry for leaving this dangling for so long!  Will close and eventually implement "all calls are with a non-NULL pe"

**Pull request closed without merging**